src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,283 @@
+         */
+        void build_maps();
+
+        /// Check if one of the statements assigns this variable (i.e. \f x' = f(x, y, x) \f) and is
+        /// still tagged
+        inline bool is_var_assigned_here(const std::string& var) const {
+            const auto it = var2statement_.find(var);
+            return it != var2statement_.end() && tags_[it->second];
+        }
+
+        /// Check if all the statements found their position
+        inline bool is_all_untagged() const {
+            return std::find(tags_.begin(), tags_.end(), true) == tags_.end();
+        }
+        /**
+         * \brief Look for \p var in \ref var2statement_ and emplace back that statement in \p
+         * new_statements
+         *
+         * If there is no \p var key in \ref var2statement_, return
+         */
+        bool try_emplace_back_statement(ast::StatementVector& new_statements,
+                                        const std::string& var);
+
+
+        /// Emplace back the first statement in \ref statements that is marked for updating in \ref
+        /// tags_
+        bool emplace_back_next_statement(ast::StatementVector& new_statements);
+
+        /// Emplace back all the statements that are marked for updating in \ref tags_
+        size_t emplace_back_all_statements(ast::StatementVector& new_statements,
+                                           const bool is_logger = false);
+
+        /**
+         * \brief Tag all the statements that depend on \p var for updating
+         *
+         * This is necessary when an assignment has invalidated this variable
+         */
+        size_t tag_dependant_statements(const std::string& var);
+
+        /// Mark that all the statements need updating (probably unused)
+        void tag_all_statements();
+
+        // TODO remove
+        void print();
+
+        /**
+         * \brief var -> (depends on) vars
+         *
+         * The statements are assignments. Thus, in general, an assigned variable depends on
+         * the variables in the rhs of the assignment operator. This map keeps track of this
+         * dependency
+         */
+        std::unordered_map<std::string, std::unordered_set<std::string>> dependency_map_;
+
+        /**
+         * \brief var -> (statements that depend on) statements
+         *
+         * This the "reverse" of \ref dependency_map_. Given a certain variable it provides
+         * the statements that depend on it. It is a set because we want to print them in
+         * order and we do not want duplicates. The value is the index in \ref statements_ or \ref
+         * tags_
+         */
+        std::unordered_map<std::string, std::set<size_t>> var2dependants_;
+
+        /**
+         * \brief var -> statement that sets that var
+         *
+         * Given a certain variable we get the statement that sets that variable as index in \ref
+         * statements_ or \ref tags_
+         */
+        std::unordered_map<std::string, size_t> var2statement_;
```

Comment:
ususally this is used to find out what is the solution statement that assigns that variable. That is different from a binary expression in the original code that assigns the variable. I am afraid this would confuse some. In addition, there is a statement_ member in SolutionSorter and this umap points at that